### PR TITLE
Improved delimiter lexing

### PR DESCRIPTION
**Pull request by hamishknight**
Refactor the delimiter lexing logic, and change the behavior such that:

- We now diagnose unprintable ASCII characters to match behavior with the C++ lexer.
- We now allow the C++ lexer to better recover from a missing closing delimiter.
- We now support `rx'...'` for experimental syntax.

Additionally, implement a heuristic that allows skipping over single quotes in `re'...'` literals if the preceding characters are `(?`, `(?(`, `\k`, `\g` or `(?C`.  These would not be valid literal endings anyway, and this allows us to support their single-quoted syntax.

This should be able to integrate without any C++ side changes, though I have some extra test cases I want to commit there whenever this gets integrated in.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

File: Sources/_MatchingEngine/Regex/Parse/DelimiterLexing.swift [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+// TODO: mock up multi-line soon
+
+enum Delimiter: Hashable, CaseIterable {
+  case traditional
+  case experimental
+  case reSingleQuote
```
Comment (milseman) [inline Sources/_MatchingEngine/Regex/Parse/DelimiterLexing.swift:17]: How about `rxSingleQuote` for experimental?

File: Sources/_MatchingEngine/Regex/Parse/DelimiterLexing.swift [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    switch kind {
+    case .endOfString: return "unterminated regex literal"
+    case .invalidUTF8: return "invalid UTF-8 found in source file"
+    case .unknownDelimiter: return "unknown regex literal delimiter"
+    case .unprintableASCII: return "unprintable ASCII character found in source file"
+    }
+  }
+}
+
+fileprivate struct DelimiterLexer {
```
Comment (milseman) [inline Sources/_MatchingEngine/Regex/Parse/DelimiterLexing.swift:68]: In the end, we might want something that can feed `Source` (or even replace `Source`)

**Review by milseman — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please test

File: Sources/_MatchingEngine/Regex/Parse/DelimiterLexing.swift (hunk 2 above)
Comment (hamishknight) [reply Sources/_MatchingEngine/Regex/Parse/DelimiterLexing.swift:70]: Yeah would be nice to unify on an implementation that scans and produces unicode scalars for both

**Pull request merged**